### 1. Подберите скорость обучения (alpha) и количество итераций

In [45]:
import numpy as np
import matplotlib.pyplot as plt

In [46]:
y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60]

In [47]:
X = np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
              [1, 1, 2, 1, 3, 0, 5, 10, 1, 2]])

In [48]:
n = X.shape[1]
alpha = 1e-2
W = np.array([1, 0.5])
min_err_dist = 1e-6 
W, alpha

(array([1. , 0.5]), 0.01)

In [49]:
y = [45, 55, 50, 55, 60, 35, 75, 80, 50, 60]

In [50]:
def calc_mse(y, y_pred):
    err = np.mean((y - y_pred)**2)
    return err

In [58]:
def test_alpha_niter(y, y_pred, alpha, max_iter, min_err_dist):
    W = np.array([1, 0.5])
    err = np.inf
    for i in range(max_iter):
        y_pred = np.dot(W, X)
        err_new = calc_mse(y, y_pred)
        if abs(err_new) > 100000000:
            print(f'INF!!! Iteration {i}: W_new = {W}, MSE = {round(err,8)}')
            break
        if abs(err_new - err) < min_weight_dist:
            print(f'Iteration {i}: W_new = {W}, MSE = {round(err,8)}')
            break
        err = err_new
        for ii in range(W.shape[0]):
            W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
            # if i % 100 == 0:
            #    print(i, W, err)

In [60]:
test_alpha_niter(y, y_pred, 0.01, 100000, min_err_dist)

Iteration 807: W_new = [46.17687059  3.968564  ], MSE = 36.88142558


In [61]:
test_alpha_niter(y, y_pred, 0.001, 100000, min_err_dist)

Iteration 6992: W_new = [46.15635543  3.97235228], MSE = 36.88185819


In [62]:
test_alpha_niter(y, y_pred, 0.1, 10000, min_err_dist)

INF!!! Iteration 8: W_new = [ -547.21435461 -3109.3221263 ], MSE = 37268269.84065965


In [63]:
test_alpha_niter(y, y_pred, 0.02, 100000, min_err_dist)

Iteration 419: W_new = [46.17976523  3.96802949], MSE = 36.88140094


In [64]:
test_alpha_niter(y, y_pred, 0.05, 100000, min_err_dist)

Iteration 174: W_new = [46.18224549  3.96757149], MSE = 36.88138702


In [66]:
test_alpha_niter(y, y_pred, 0.06, 100000, min_err_dist)

Iteration 146: W_new = [46.18267874  3.96749149], MSE = 36.88138524


In [67]:
test_alpha_niter(y, y_pred, 0.07, 100000, min_err_dist)

INF!!! Iteration 52: W_new = [ -462.34513109 -2745.20692215], MSE = 95460529.1595758


In [68]:
test_alpha_niter(y, y_pred, 0.065, 100000, min_err_dist)

Iteration 235: W_new = [46.18638055  3.96769928], MSE = 36.88139011


Наименьшее число итераций (146) и наименьшее значение MSE получилось при $aplha = 0.06$

### 2 (опция). В этом коде мы избавляемся от итераций по весам, но тут есть ошибка, исправьте ее:

In [73]:
n = X.shape[1]
alpha = 1e-2
W = np.array([1, 0.5])


for i in range(1000):
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)
    #for ii in range(W.shape[0]):
    #    W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
    W -= (alpha * (1/n * 2 * (X @ (y_pred - y))))
    if i % 100 == 0:
        print(i, W, err)


0 [2.084 3.862] 3068.75
100 [31.19721117  6.73467271] 160.2197550756897
200 [40.91589024  4.94004427] 52.12992440539934
300 [44.33310564  4.30902931] 38.76658299831662
400 [45.53464353  4.08715614] 37.11444891266922
500 [45.95712005  4.00914262] 36.91019258531745
600 [46.10566836  3.98171204] 36.88494000586543
700 [46.15789989  3.97206709] 36.88181798376089
800 [46.17626519  3.9686758 ] 36.88143200251729
900 [46.18272267  3.96748337] 36.88138428295753


### 3 (опция). Реализовать один из критериев останова, перечисленный в методичке.

In [88]:
def test_alpha_niter(y, y_pred, alpha, max_iter, min_err_dist, min_weight_dist):
    W = np.array([1, 0.5])
    W_new = np.array([1000, 0.5])
    err = np.inf
    for i in range(max_iter):
        y_pred = np.dot(W, X)
        err_new = calc_mse(y, y_pred)
        
        if abs(err_new) > 100000000:
            print('Процесс разошелся')
            print(f'INF!!! Iteration {i}: W_new = {W_new}, MSE = {round(err,8)}')
            break
        if abs(err_new - err) < min_err_dist:
            print('Сработал критерий останова: изменение MSE меньше заданного числа')
            print(f'Iteration {i}: W_new = {W_new}, MSE = {round(err,8)}')
            break
        
        err = err_new
        for ii in range(W.shape[0]):
            W_new[ii] = W[ii] - alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
            # if i % 100 == 0:
            #    print(i, W, err)
        if np.linalg.norm(W_new - W, ord=2) < min_weight_dist:
            print('Сработал критерий останова: изменение весов меньше заданного числа')
            print(f'Iteration {i}: W_new = {W_new}, MSE = {round(err,8)}')
            break
        W = W_new.copy()
       

In [89]:
test_alpha_niter(y, y_pred, 0.06, 100000, 1e-6, 1e-6)

Сработал критерий останова: изменение MSE меньше заданного числа
Iteration 146: W_new = [46.18267874  3.96749149], MSE = 36.88138524


In [93]:
test_alpha_niter(y, y_pred, 0.06, 100000, 1e-6, 1e-2)

Сработал критерий останова: изменение весов меньше заданного числа
Iteration 88: W_new = [46.0467942   3.99258369], MSE = 36.89326642
